src/DotNetty.Handlers/Streams/ChunkedWriteHandler.cs

```diff
@@ -228,54 +229,62 @@ void DoFlush(IChannelHandlerContext context)
                         // be closed before its not written.
                         //
                         // See https://github.com/netty/netty/issues/303
-                        future.ContinueWith((_, state) =>
+                        CloseOnComplete(writeFuture, current, chunks);
+
+                        async void CloseOnComplete(ValueTask future, PendingWrite promise, IChunkedInput<T> input)
+                        {
+                            try
                             {
-                                var pendingTask = (PendingWrite)state;
-                                CloseInput((IChunkedInput<T>)pendingTask.Message);
-                                pendingTask.Success();
-                            },
-                            current, 
-                            TaskContinuationOptions.ExecuteSynchronously);
+                                await future;
+                            }
+                            finally
+                            {
+                                promise.Progress(input.Progress, input.Length);
+                                promise.TryComplete();
+                                CloseInput(input);
+                            }
+                        }
                     }
                     else if (channel.IsWritable)
                     {
-                        future.ContinueWith((task, state) =>
+                        ProgressOnComplete(writeFuture, current, chunks);
+                        
+                        async void ProgressOnComplete(ValueTask future, PendingWrite promise, IChunkedInput<T> input)
+                        {
+                            try
                             {
-                                var pendingTask = (PendingWrite)state;
-                                if (task.IsFaulted)
-                                {
-                                    CloseInput((IChunkedInput<T>)pendingTask.Message);
-                                    pendingTask.Fail(task.Exception);
-                                }
-                                else
-                                {
-                                    pendingTask.Progress(chunks.Progress, chunks.Length);
-                                }
-                            },
-                            current,
-                            TaskContinuationOptions.ExecuteSynchronously);
+                                await future;
+                                promise.Progress(input.Progress, input.Length);
+                            }
+                            catch(Exception ex)
+                            {
+                                CloseInput((IChunkedInput<T>)promise.Message);
+                                promise.TrySetException(ex);
+                            }
+                        }
```

Comment:
@StormHub , I've aligned behavior with netty here: progress() never completes promise
https://github.com/netty/netty/blob/8d78893a76a8c14c18b304ffe1fce2dc43ed4206/handler/src/main/java/io/netty/handler/stream/ChunkedWriteHandler.java#L278

Any reason it was ported in different way?

Reply:
@maksimkim I don't think there is a particular reason. Let me double check.

Reply:
@maksimkim I think I did it wrong. Netty progress never completes the promise.

Reply:
@maksimkim also note that success also calls progress
https://github.com/netty/netty/blob/4f982be91e6d61c7c7a9d8d63115e0ebe65b2774/handler/src/main/java/io/netty/handler/stream/ChunkedWriteHandler.java#L350